_posts/drafts/hack-go-type-system.md

```diff
@@ -0,0 +1,842 @@
+	0x003e 00062 (cast.go:7)	LEAQ	"".autotmp_0+72(SP), CX
+	0x0043 00067 (cast.go:7)	MOVQ	CX, 8(SP)
+	0x0048 00072 (cast.go:7)	LEAQ	"".autotmp_3+48(SP), CX
+	0x004d 00077 (cast.go:7)	MOVQ	CX, 16(SP)
+	0x0052 00082 (cast.go:7)	PCDATA	$0, $0
+	0x0052 00082 (cast.go:7)	CALL	runtime.convT2E(SB)
+	0x0057 00087 (cast.go:7)	MOVQ	32(SP), AX
+	0x005c 00092 (cast.go:7)	MOVQ	24(SP), CX
+```
+
+My first move is on **runtime.convT2E(SB)**:
+
+```	
+func convT2E(t *_type, elem unsafe.Pointer, x unsafe.Pointer) (e eface) {
+	if raceenabled {
+		raceReadObjectPC(t, elem, getcallerpc(unsafe.Pointer(&t)), funcPC(convT2E))
+	}
+	if msanenabled {
+		msanread(elem, t.size)
+	}
+	if isDirectIface(t) {
+		throw("direct convT2E")
+	}
+	if x == nil {
+		x = newobject(t)
+		// TODO: We allocate a zeroed object only to overwrite it with
+		// actual data. Figure out how to avoid zeroing. Also below in convT2I.
+	}
+	typedmemmove(t, x, elem)
+	e._type = t
+	e.data = x
+	return
+}
+```
+
+Yeah, definitely seems to be initializing the eface and returning it.
+Now I just have to figure a way to get my hands on the **_type** pointer.
+
+I still don't have a good idea on how to get the *_type, or how to manipulate
+the eface type. My guess would be to just cast it as a pointer and do some
+old school pointer manipulation, but I'm not sure yet.
+
+One function that is a good candidate to give some pointers
+on how to do it is **reflect.TypeOf**:
+
+```
+func TypeOf(i interface{}) Type {
+	eface := *(*emptyInterface)(unsafe.Pointer(&i))
+	return toType(eface.typ)
+}
+```
+
+Yeah, just cast the pointer to a eface pointer:
+
+```
+// emptyInterface is the header for an interface{} value.
+type emptyInterface struct {
+	typ  *rtype
+	word unsafe.Pointer
+}
+```
+
+It seems that although the eface was private on the **runtime**
+package it is copied here on the **reflect** package. Well, if the
+reflect package can do it, so can I :-).
+
+Before going on, I was curious about where the types are initialized.
+It seems that there is just one unique pointer with all the type
+information for each type. Thanks to [vim-go](https://github.com/fatih/vim-go)
+and [go guru](https://godoc.org/golang.org/x/tools/cmd/guru)
+for the invaluable help on analysing code and allowing me to
+check all the referers to a type. Thanks to these tools it has been
+pretty easy to find this on symtab:
+
+```
+// moduledata records information about the layout of the executable
+// image. It is written by the linker. Any changes here must be
+// matched changes to the code in cmd/internal/ld/symtab.go:symtab.
+// moduledata is stored in read-only memory; none of the pointers here
+// are visible to the garbage collector.
+type moduledata struct {
+	pclntable    []byte
+	ftab         []functab
+	filetab      []uint32
+	findfunctab  uintptr
+	minpc, maxpc uintptr
+
+	text, etext           uintptr
+	noptrdata, enoptrdata uintptr
+	data, edata           uintptr
+	bss, ebss             uintptr
```

Comment:
https://en.wikipedia.org/wiki/.bss

Block Started by Symbol. Never remember the acronym =P, but it is where static variables live on a binary.